File: modules/schema/call_api.py
from pydantic import BaseModel, field_validator
from typing import Optional, List, Any, Literal


class UserCard(BaseModel):
    model_config = {"extra": "ignore"}
    cardId: Optional[int] = None
    defaultImage: Optional[Literal["original", "special_training"]] = None
    level: Optional[int] = None
    masterRank: Optional[int] = None
    specialTrainingStatus: Optional[Literal["not_doing", "done"]] = None


class UserCheerfulCarnival(BaseModel):
    model_config = {"extra": "ignore"}
    cheerfulCarnivalTeamId: Optional[int] = None
    eventId: Optional[int] = None
    registerAt: Optional[int] = None
    teamChangeCount: Optional[int] = None


class UserProfile(BaseModel):
    model_config = {"extra": "ignore"}
    profileImageType: Optional[str] = None
    twitterId: Optional[str] = None
    userId: Optional[int] = None
    word: Optional[str] = None


class UserProfileHonor(BaseModel):
    model_config = {"extra": "ignore"}
    bondsHonorViewType: Optional[
        Literal["none", "normal", "reverse", "normal_unit_virtual_singer", "reverse_unit_virtual_singer"]
    ] = None
    bondsHonorWordId: Optional[int] = None
    honorId: Optional[int] = None
    honorLevel: Optional[int] = None
    profileHonorType: Optional[
        Literal[
            "normal",
            "bonds",
        ]
    ] = None
    seq: Optional[int] = None


class PlayerRankingSchema(BaseModel):
    model_config = {"extra": "ignore"}
    isOwn: Optional[bool] = None
    name: Optional[str] = None
    rank: Optional[int] = None
    score: Optional[int] = None
    userCard: Optional[UserCard] = None
    userCheerfulCarnival: Optional[UserCheerfulCarnival] = None
    userHonorMissions: Optional[List[Any]] = None
    userId: Optional[int] = None
    userProfile: Optional[UserProfile] = None
    userProfileHonors: Optional[List[UserProfileHonor]] = None

    @field_validator("userCheerfulCarnival", mode="before")
    @classmethod
    def handle_empty_user_cheerful_carnival(cls, v: Any) -> Any:
        if isinstance(v, dict) and not v:
            return None
        return v

    @field_validator("userProfileHonors", mode="before")
    @classmethod
    def handle_empty_user_profile_honors(cls, v: Any) -> Any:
        if isinstance(v, list) and not v:
            return None
        return v


class UserWorldBloomChapterRankingBase(BaseModel):
    model_config = {"extra": "ignore"}
    eventId: Optional[int] = None
    gameCharacterId: Optional[int] = None
    isWorldBloomChapterAggregate: Optional[bool] = None


class UserWorldBloomChapterRanking(UserWorldBloomChapterRankingBase):
    rankings: Optional[List[PlayerRankingSchema]] = None


class UserWorldBloomChapterRankingBorder(UserWorldBloomChapterRankingBase):
    borderRankings: Optional[List[PlayerRankingSchema]] = None


class Top100RankingResponse(BaseModel):
    model_config = {"extra": "ignore"}
    isEventAggregate: Optional[bool] = None
    rankings: Optional[List[PlayerRankingSchema]] = None
    userRankingStatus: Literal["normal", "event_aggregate", "world_aggregate"] = "normal"
    userWorldBloomChapterRankings: Optional[List[UserWorldBloomChapterRanking]] = None


class BorderRankingResponse(BaseModel):
    model_config = {"extra": "ignore"}
    eventId: Optional[int] = None
    isEventAggregate: Optional[bool] = None
    borderRankings: Optional[List[PlayerRankingSchema]] = None
    userWorldBloomChapterRankingBorders: Optional[List[UserWorldBloomChapterRankingBorder]] = None

File: utils/model/callapi.go
package model

type UserCard struct {
	CardID                *int    `json:"cardId,omitempty"`
	DefaultImage          *string `json:"defaultImage,omitempty"`
	Level                 *int    `json:"level,omitempty"`
	MasterRank            *int    `json:"masterRank,omitempty"`
	SpecialTrainingStatus *string `json:"specialTrainingStatus,omitempty"`
}

type UserCheerfulCarnival struct {
	CheerfulCarnivalTeamID *int   `json:"cheerfulCarnivalTeamId,omitempty"`
	EventID                *int   `json:"eventId,omitempty"`
	RegisterAt             *int64 `json:"registerAt,omitempty"`
	TeamChangeCount        *int   `json:"teamChangeCount,omitempty"`
}

type UserProfile struct {
	ProfileImageType *string `json:"profileImageType,omitempty"`
	TwitterID        *string `json:"twitterId,omitempty"`
	UserID           *int    `json:"userId,omitempty"`
	Word             *string `json:"word,omitempty"`
}

type UserProfileHonor struct {
	BondsHonorViewType *string `json:"bondsHonorViewType,omitempty"`
	BondsHonorWordID   *int    `json:"bondsHonorWordId,omitempty"`
	HonorID            *int    `json:"honorId,omitempty"`
	HonorLevel         *int    `json:"honorLevel,omitempty"`
	ProfileHonorType   *string `json:"profileHonorType,omitempty"`
	Seq                *int    `json:"seq,omitempty"`
}

type PlayerRankingSchema struct {
	IsOwn                *bool                 `json:"isOwn,omitempty"`
	Name                 *string               `json:"name,omitempty"`
	Rank                 *int                  `json:"rank,omitempty"`
	Score                *int                  `json:"score,omitempty"`
	UserCard             *UserCard             `json:"userCard,omitempty"`
	UserCheerfulCarnival *UserCheerfulCarnival `json:"userCheerfulCarnival,omitempty"`
	UserHonorMissions    []interface{}         `json:"userHonorMissions,omitempty"`
	UserID               *int                  `json:"userId,omitempty"`
	UserProfile          *UserProfile          `json:"userProfile,omitempty"`
	UserProfileHonors    []UserProfileHonor    `json:"userProfileHonors,omitempty"`
}

type UserWorldBloomChapterRankingBase struct {
	EventID                      *int  `json:"eventId,omitempty"`
	GameCharacterID              *int  `json:"gameCharacterId,omitempty"`
	IsWorldBloomChapterAggregate *bool `json:"isWorldBloomChapterAggregate,omitempty"`
}

type UserWorldBloomChapterRanking struct {
	UserWorldBloomChapterRankingBase
	Rankings []PlayerRankingSchema `json:"rankings,omitempty"`
}

type UserWorldBloomChapterRankingBorder struct {
	UserWorldBloomChapterRankingBase
	BorderRankings []PlayerRankingSchema `json:"borderRankings,omitempty"`
}

type Top100RankingResponse struct {
	IsEventAggregate              *bool                          `json:"isEventAggregate,omitempty"`
	Rankings                      []PlayerRankingSchema          `json:"rankings,omitempty"`
	UserRankingStatus             string                         `json:"userRankingStatus"`
	UserWorldBloomChapterRankings []UserWorldBloomChapterRanking `json:"userWorldBloomChapterRankings,omitempty"`
}

type BorderRankingResponse struct {
	EventID                             *int                                 `json:"eventId,omitempty"`
	IsEventAggregate                    *bool                                `json:"isEventAggregate,omitempty"`
	BorderRankings                      []PlayerRankingSchema                `json:"borderRankings,omitempty"`
	UserWorldBloomChapterRankingBorders []UserWorldBloomChapterRankingBorder `json:"userWorldBloomChapterRankingBorders,omitempty"`
}
